src/WholeBodyControllers/src/QPInverseKinematics.cpp

```diff
@@ -546,14 +677,38 @@ void WalkingQPIK::evaluateGradientVector()
 
     // Neck orientation
     iDynTree::Matrix3x3 errorNeckAttitude = iDynTreeUtilities::Rotation::skewSymmetric(m_neckOrientation * m_desiredNeckOrientation.inverse());
-    gradient = -neckJacobian.transpose() * m_neckWeight * (-m_kNeck
-                                                           * iDynTree::unskew(iDynTree::toEigen(errorNeckAttitude)));
 
-    // if the hand retargeting is not enable the regularization term should be used as usual
-    if(!m_enableHandRetargeting)
-       gradient += -jointRegularizationGradient * jointRegularizationGains.asDiagonal()
-           * (regularizationTerm - jointPosition);
+    if(!m_enableJointRetargeting)
+    {
+        auto jointRegularizationWeights(iDynTree::toEigen(m_jointRegularizationWeights));
+
+        gradient = -neckJacobian.transpose() * m_neckWeight * (-m_kNeck * iDynTree::unskew(iDynTree::toEigen(errorNeckAttitude)));
+
+        // g = Weight * K_p * (regularizationTerm - jointPosition)
+        // Weight  and K_p are two diagonal matrices so their product can be also evaluated multiplying component-wise
+        // the elements of the vectors and then generating the diagonal matrix
+        gradient.tail(m_actuatedDOFs) += (-jointRegularizationWeights.cwiseProduct(jointRegularizationGains)).asDiagonal()
+            * (regularizationTerm - jointPosition);
+    }
     else
+    {
+        auto jointRetargetingGains(iDynTree::toEigen(m_jointRetargetingGains));
+        auto jointRetargetingValues(iDynTree::toEigen(m_retargetingJointValue));
+
+        gradient = -neckJacobian.transpose() * m_torsoWeightSmoother->getPos()(0)
+            * (-m_kNeck * iDynTree::unskew(iDynTree::toEigen(errorNeckAttitude)));
+
+        // g = Weight * K_p * (regularizationTerm - jointPosition)
+        // Weight  and K_p are two diagonal matrices so their product can be also evaluated multiplying component-wise
+        // the elements of the vectors and then generating the diagonal matrix
+        gradient.tail(m_actuatedDOFs) += (-jointRegularizationGains.cwiseProduct(
+                                              iDynTree::toEigen(m_jointRegularizationWeightSmoother->getPos()))).asDiagonal()
+            * (regularizationTerm - jointPosition)
+            + (-jointRetargetingGains.cwiseProduct(iDynTree::toEigen(m_jointRetargetingWeightSmoother->getPos()))).asDiagonal()
+            * (jointRetargetingValues - jointPosition);
```

Comment:
When you say splitting do you mean to store half of the computation in an `Eigen::VectorXd`? I

Reply:
Yes, you can use some intermediate buffers. Also if you have something like
``
a = (b*c) + (c*d);
``
you can split it in
``
a = (b*c);
a += (c*d);
``
For more info, https://eigen.tuxfamily.org/dox/TopicLazyEvaluation.html

Reply:
> Yes, you can use some intermediate buffers. Also if you have something like
> 
> ```
> a = (b*c) + (d*e);
> ```
> 
> you can split it in
> 
> ```
> a = (b*c);
> a += (d*e);
> ```
> 
> For more info, https://eigen.tuxfamily.org/dox/TopicLazyEvaluation.html

Shouldn't this reduce the performances? As far as I understood `a = (b*c) + (d*e)` Eigen will decide whether to evaluate it into a temporary variable. Thus if we are lucky no temporary variables are required. On the other hand 
```
a = (b*c);
a += (d*e);
```
will require the evaluation of the first part in the variable `a`. 

Probably this may solve our problems
```cpp
auto tmp1 = (b*c);
auto tmp2 = (d*e);
a = tmp1 + tmp2;
```
however, I am not a big fan of using `auto` with `Eigen` (you never know which kind of type are you using)

By the way, I completely agree that sometimes the readability is more important than the performances, so I can split the computation in two parts.

Reply:
> Shouldn't this reduce the performances? As far as I understood a = (b*c) + (d*e) Eigen will decide whether to evaluate it into a temporary variable. Thus if we are lucky no temporary variables are required.

Matrix multiplication is one of the cases where Eigen evaluates the sub-expression in a temporary object. Hence, I guess that the advantage of avoiding an assignment is eaten up by the need for allocating memory.

Reply:
Thanks for the explanation. 
f7442cff90563d0294c24214332117865ad41652 I split the computation as you suggested